driver-core/src/main/com/mongodb/internal/connection/SocksSocket.java

```diff
@@ -0,0 +1,444 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.connection;
+
+import com.mongodb.connection.ProxySettings;
+import com.mongodb.internal.Timeout;
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.lang.Nullable;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ConnectException;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.assertTrue;
+import static com.mongodb.assertions.Assertions.fail;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public final class SocksSocket extends Socket {
+    private static final Pattern DOMAIN_PATTERN = Pattern.compile("^(([a-zA-Z0-9]([a-zA-Z0-9\\-]{0,61}[a-zA-Z0-9])?\\.)+[a-zA-Z]{2,6}|localhost)$");
+    private static final byte LENGTH_OF_IPV4 = 4;
+    private static final byte LENGTH_OF_IPV6 = 16;
+    private static final byte SOCKS_VERSION = 5;
+    private static final byte RESERVED = 0x00;
+    private static final byte PORT_SIZE = 2;
+    private static final byte AUTHENTICATION_SUCCEEDED_STATUS = 0x00;
+    private static final byte REQUEST_OK = 0;
+    private static final byte ADDRESS_TYPE_DOMAIN_NAME = 3;
+    private static final byte ADDRESS_TYPE_IPV4 = 1;
+    private static final byte ADDRESS_TYPE_IPV6 = 4;
+    private static final int DEFAULT_PORT = 1080;
+    public static final String IP_PARSING_ERROR_SUFFIX = " is not an IP string literal";
+    private final SocksAuthenticationMethod[] authenticationMethods;
+    private final InetSocketAddress proxyAddress;
+    private InetSocketAddress remoteAddress;
+    @Nullable
+    private String proxyUsername;
+    @Nullable
+    private String proxyPassword;
+    @Nullable
+    private final Socket socket;
+    private InputStream inputStream;
+    private OutputStream outputStream;
+
+    public SocksSocket(final ProxySettings proxySettings) {
+        this(null, proxySettings);
+    }
+
+    public SocksSocket(@Nullable final Socket socket, final ProxySettings proxySettings) {
```

Comment:
After our verbal discussion, I'm sharing the outcome below.

The behavior stems from a specific logic in the code flow. Specifically, the` initializeSocketOverSocksProxy()` method follows these steps:

- It creates a basic TCP socket using the socketFactory.
- Then, it customizes this socket with some settings using configureSocket(createdSocket, settings).
- Afterward, it wraps the socket with extra functionality in a SocksSocket instance.

```
 private Socket initializeSocketOverSocksProxy() throws IOException {
        Socket createdSocket = socketFactory.createSocket();
        configureSocket(createdSocket, settings);
        SocksSocket socksProxy = new SocksSocket(createdSocket, settings.getProxySettings());

        socksProxy.connect(InetSocketAddress.createUnresolved(address.getHost(), address.getPort()),
                settings.getConnectTimeout(TimeUnit.MILLISECONDS));
        return socksProxy;
    }
```
   There was a reasonable question as to why we don't just give the `socksProxy` socket directly to the configure part, like this: 
   
```
        Socket createdSocket = socketFactory.createSocket();
        SocksSocket socksProxy = new SocksSocket(createdSocket, settings.getProxySettings());
        configureSocket(socksProxy);
```

The issue arises because, in the `configureSocket` method, we make use of the [setExtendedSocketOptions](https://github.com/mongodb/mongo-java-driver/blob/311515e2605527368b2c7fd613c821fc81527303/driver-core/src/main/com/mongodb/internal/connection/SocketStreamHelper.java#L95) method. This method configures the socket with specific methods from Java 11 via reflection due to our commitment to supporting Java 8 as the minimum version.

However, in `SocksSocket`, we can't easily add these methods during compile time to pass on their execution to the wrapped socket. This limitation prevents us from directly extending those methods. As a result, our current approach involves configuring the newly created socket first. After this configuration step, we proceed to wrap it in the 'SocksSocket', introducing the additional necessary logic.